### PR TITLE
Opt sat

### PR DESCRIPTION
**Pull request by vhavlena**
This PR deals with underapproximation of string constraints

### CODE REVIEW

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 1]
```diff
@@ -2062,9 +2116,10 @@ namespace smt::noodler {
      * @param len_formula Formula to be check
      * @return lbool Sat
      */
-    lbool theory_str_noodler::check_len_sat(expr_ref len_formula) {
+    lbool theory_str_noodler::check_len_sat(expr_ref len_formula, model_ref &mod) {
         int_expr_solver m_int_solver(get_manager(), get_context().get_fparams());
         m_int_solver.initialize(get_context());
-        return m_int_solver.check_sat(len_formula);
+        auto ret = m_int_solver.check_sat(len_formula);
+        return ret;
```
Comment (Adda0) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:2123]: There is no reason for this change, is there?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 2]
```diff
@@ -781,13 +781,20 @@ namespace smt::noodler {
         expr_ref lengths(m);
         std::unordered_set<BasicTerm> init_length_sensitive_vars{ get_init_length_vars(aut_assignment) };
 
-        DecisionProcedure dec_proc = DecisionProcedure{ instance, aut_assignment, init_length_sensitive_vars, m, m_util_s, m_util_a };
+        // use underapproximation to solve
+        if(m_params.m_underapproximation && solve_underapprox(instance, aut_assignment, init_length_sensitive_vars) == l_true) {
+            STRACE("str", tout << "underapprox sat \n";);
+            return FC_DONE;
+        }
+
```
Comment (Adda0) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:789]: Better separate underapproximation from the regular algorithm checking the whole instance.
```suggestion
        // Underapproximation is unsat. We need to check the whole instance instead.
```

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 3]
```diff
@@ -375,12 +375,12 @@ namespace smt::noodler {
...
                 
                 
-                return;
+                //return;
             }
 
             // build axiom 2: Length(a_str) == 0 <=> a_str == ""
             {
-                return;
+                // return;
```
Comment (Adda0) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:383]: I believe we can remove these returns now.

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 4]
```diff
@@ -540,7 +549,7 @@ namespace smt::noodler {
                 std::set<VarNode> occurs_act = this->formula.get_var_occurr(side[i]);
 
                 // do not include length variables
-                if(this->len_variables.find(side[i]) != this->len_variables.end()) {
+                if(false && this->len_variables.find(side[i]) != this->len_variables.end()) {
```
Comment (Adda0) [inline src/smt/theory_str_noodler/formula_preprocess.cpp:552]: We should resolve this better than passing `false` to the if statement.

File: src/smt/params/theory_str_noodler_params.cpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#include "smt/params/theory_str_noodler_params.h"
+#include "smt/params/smt_params_helper.hpp"
+
+void theory_str_noodler_params::updt_params(params_ref const & _p) {
+    smt_params_helper p(_p);
+    m_underapproximation = p.str_underapprox();
+    m_preprocess_red = p.str_preprocess_red();
+}
+
+#define DISPLAY_PARAM(X) out << #X"=" << X << std::endl;
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/params/theory_str_noodler_params.cpp:11]: Ok

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 1 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:2123]: Probably not

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 6]
```diff
@@ -2062,9 +2116,10 @@ namespace smt::noodler {
      * @param len_formula Formula to be check
      * @return lbool Sat
      */
-    lbool theory_str_noodler::check_len_sat(expr_ref len_formula) {
+    lbool theory_str_noodler::check_len_sat(expr_ref len_formula, model_ref &mod) {
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:2119]: It is prepared for a future use

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 3 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:383]: Refactoring is on my todo list. But not now

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 4 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:552]: Sure, I will change it at some point

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 2 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:789]: Great, can you push the comment there?

**Comment by Adda0** [comment]
We will keep a link to this PR somewhere when we refactor the code. For now, let us merge the PR.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 2 above)
Comment (Adda0) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:789]: (Still remains to be added.)

**Pull request merged**